jina/peapods/grpc/__init__.py

```diff
@@ -0,0 +1,151 @@
+import asyncio
+import inspect
+from typing import Optional, Callable
+
+import grpc
+from google.protobuf import struct_pb2
+
+from jina.logging.logger import JinaLogger
+from jina.proto import jina_pb2_grpc, jina_pb2
+from jina.types.message import Message
+from jina.types.message.common import ControlMessage
+from jina.types.routing.table import RoutingTable
+
+
+class Grpclet(jina_pb2_grpc.JinaDataRequestRPCServicer):
+    """A `Grpclet` object can send/receive Messages via gRPC.
+
+    :param async_loop: asyncio loop to use
+    :param args: the parsed arguments from the CLI
+    :param message_callback: the callback to call on received messages
+    :param logger: the logger to use
+    """
+
+    def __init__(
+        self,
+        args: 'argparse.Namespace',
+        message_callback: Callable[['Message'], None],
+        logger: Optional['JinaLogger'] = None,
+    ):
+        self.args = args
+        self._stubs = {}
+        self._logger = logger or JinaLogger(self.__class__.__name__)
+        self.callback = message_callback
+        self.msg_recv = 0
+        self.msg_sent = 0
+
+    async def send_message(self, msg: 'Message', **kwargs):
+        """
+        Sends a message via gRPC to the target indicated in the message's routing table
+        :param msg: the protobuf message to send
+        :param kwargs: Additional arguments.
+        """
+        routing_table = RoutingTable(msg.envelope.routing_table)
+
+        next_targets = routing_table.get_next_targets()
+
+        for target, _ in next_targets:
+            pod_address = target.active_target_pod.full_address
+
+            new_message = self._add_envelope(msg, target)
+
+            if pod_address not in self._stubs:
+                self._stubs[pod_address] = Grpclet._create_grpc_stub(pod_address)
+
+            try:
+                self.msg_sent += 1
+
+                # this wraps the awaitable object from grpc as a coroutine so it can be used as a task
+                # the grpc call function is not a coroutine but some _AioCall
+                async def task_wrapper(new_message, pod_address):
+                    await self._stubs[pod_address].Call(new_message)
+
+                asyncio.create_task(task_wrapper(new_message, pod_address))
+            except grpc.RpcError as ex:
+                self._logger.error('Sending data request via grpc failed', ex)
+                raise ex
+
+    @staticmethod
+    def send_ctrl_msg(pod_address: str, command: str):
+        """
+        Sends a control message via gRPC to pod_address
+        :param pod_address: the pod to send the command to
+        :param command: the command to send (TERMINATE/ACTIVATE/...)
+        :returns: Empty protobuf struct
+        """
+        stub = Grpclet._create_grpc_stub(pod_address, is_async=False)
```

Comment:
but this is a static method so I dont really want to keep state here? Also the control message are very rare.
Also in the case of the ready check it will make not much of a difference. While its running in the loop all grpc calls will fail as there is no connection. Once everything works, the loop stops